packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -672,13 +748,251 @@ class ZoomPageTransitionsBuilder extends PageTransitionsBuilder {
     return _ZoomPageTransition(
       animation: animation,
       secondaryAnimation: secondaryAnimation,
-      allowSnapshotting: allowSnapshotting && (route?.allowSnapshotting ?? true),
+      allowSnapshotting: allowSnapshotting && route.allowSnapshotting,
       allowEnterRouteSnapshotting: allowEnterRouteSnapshotting,
       child: child,
     );
   }
 }
 
+typedef AndroidBackGestureTransitionBuilder<T> = Widget Function(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+    AndroidBackEvent? startBackEvent,
+    AndroidBackEvent? currentBackEvent);
+
+class AndroidBackGesturePageTransitionsBuilder extends PageTransitionsBuilder {
+  const AndroidBackGesturePageTransitionsBuilder({
+    this.parent = const ZoomPageTransitionsBuilder(),
+    this.backGestureTransitionBuilder = defaultBackGestureTransitionBuilder,
+  });
+
+  final PageTransitionsBuilder parent;
+  final AndroidBackGestureTransitionBuilder<dynamic>
+      backGestureTransitionBuilder;
+
+  @override
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return AndroidBackGestureDetector(
+      controller: route.controller!, // protected access
```

Comment:
Hmm this is a red flag. I think driving the route's animation in a widget like this is probably not intended...

For the record, TransitionRoute drives its animations in places like this: https://github.com/flutter/flutter/blob/84b5e799d3de03081ad0bcf770e84204020df8a0/packages/flutter/lib/src/widgets/routes.dart#L295

I wonder if you could somehow listen for the back gesture messages and update the controller from within the route?

Reply:
Any thoughts on this?